traffic_router/core/src/test/java/org/apache/traffic_control/traffic_router/core/external/BufferedResponseTest.java

```diff
@@ -0,0 +1,183 @@
+		paths.add("http://localhost:3333/crs/locations/caches");
+		paths.add("http://localhost:3333/crs/consistenthash/deliveryservice?deliveryServiceId=csd-target-1&requestPath=/");
+
+		for (final String path : paths) {
+			CloseableHttpResponse response = null;
+
+			try {
+				final HttpGet httpGet = new HttpGet(path);
+				response = httpClient.execute(httpGet);
+
+				final ObjectMapper objectMapper = new ObjectMapper(new JsonFactory());
+				final String json = EntityUtils.toString(response.getEntity());
+				final Header contentLengthHeader = response.getFirstHeader("Content-Length");
+
+				/* If the content length is too low and cuts off the response
+				 * body, objectMapper.readTree(json) will likely throw a
+				 * JsonProcessingException.
+				 */
+				objectMapper.readTree(json);
+				assertThat(response.getFirstHeader("Transfer-Encoding"), nullValue());
+				assertThat(contentLengthHeader, notNullValue());
+				assertThat(Integer.parseInt(contentLengthHeader.getValue()), equalTo(json.length()));
+			} finally {
+				if (response != null) response.close();
+			}
+		}
+	}
+
+	@Test
+	public void itSetsContentLengthHeaderForDeliveryServiceSteering() throws IOException {
+		final String testHostName = "tr.client-steering-test-1.thecdn.example.com";
+		final RequestConfig config = RequestConfig.custom().setRedirectsEnabled(false).build();
+		final List<String> paths = new ArrayList<>();
+		paths.add("/qwerytuiop/asdfghjkl?fakeClientIpAddress=12.34.56.78");
+		paths.add("/qwerytuiop/asdfghjkl?fakeClientIpAddress=12.34.56.78&format=json");
+		paths.add("/qwerytuiop/asdfghjkl?fakeClientIpAddress=12.34.56.78&" + RouterFilter.REDIRECT_QUERY_PARAM + "=false");
+		paths.add("/qwerytuiop/asdfghjkl?fakeClientIpAddress=12.34.56.78&" + RouterFilter.REDIRECT_QUERY_PARAM + "=true");
+
+
+		for (final String path : paths) {
+			final List<HttpRequestBase> requests = new ArrayList<>();
+			requests.add(new HttpHead("http://localhost:" + routerHttpPort + path));
+			requests.add(new HttpGet("http://localhost:" + routerHttpPort + path));
+
+			final List<Integer> contentLengths = new ArrayList<>();
+
+			for (final HttpRequestBase request : requests) {
+				CloseableHttpResponse response = null;
+				request.setConfig(config);
+				request.addHeader("Host", testHostName);
+
+				try {
+					response = httpClient.execute(request);
+					final Header contentLengthHeader = response.getFirstHeader("Content-Length");
+
+					assertThat(response.getFirstHeader("Transfer-Encoding"), nullValue());
+					assertThat(contentLengthHeader, notNullValue());
+					contentLengths.add(Integer.parseInt(contentLengthHeader.getValue()));
+				} finally {
+					if (response != null) response.close();
+				}
+			}
+
+			assertThat(contentLengths.size(), equalTo(2));
+			assertThat("Expected HEAD and GET requests for " + testHostName + path + "to have the same Content-Length", contentLengths.get(0), equalTo(contentLengths.get(1)));
```

Comment:
if TR's response isn't as expected, this could throw an `IndexOutOfBoundsException`. Idk much about Java testing, but you have a `try` block to catch something above, so is that alright? It also doesn't seem like an `IOException` so I'm not sure why the compiler isn't mad

Reply:
> if TR's response isn't as expected, this could throw an `IndexOutOfBoundsException`.

No, because the line above it is
https://github.com/apache/trafficcontrol/blob/81bc50d2d0c32d7fe87c56fb33a23732f46c430a/traffic_router/core/src/test/java/org/apache/traffic_control/traffic_router/core/external/BufferedResponseTest.java#L170

> Idk much about Java testing, but you have a `try` block to catch something above, so is that alright?

Is what alright?

> It also doesn't seem like an `IOException` so I'm not sure why the compiler isn't mad

[`CloseableHttpResponse.execute()`](https://github.com/apache/httpcomponents-client/blob/rel/v4.5.13/httpclient/src/main/java/org/apache/http/impl/client/CloseableHttpClient.java#L106-L107) throws `IOException`:

```java
    public CloseableHttpResponse execute(
            final HttpUriRequest request) throws IOException, ClientProtocolException {
```

Reply:
Based on what you're saying, I would guess that `assertThat` throws (some kind of `IOException`?) so that it won't actually proceed to the next line if the size assertion fails, is that right?

> Is what alright?

Is it alright for it to not catch the `IndexOutOfBoundsException`.

Reply:
> it won't actually proceed to the next line if the size assertion fails, is that right?

right